File: examples/java/basic/src/test/java/com/rpl/agent/basic/StreamSpecificAgentTest.java
package com.rpl.agent.basic;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

import com.rpl.agentorama.AgentClient;
import com.rpl.agentorama.AgentInvoke;
import com.rpl.agentorama.AgentManager;
import com.rpl.rama.test.InProcessCluster;
import com.rpl.rama.test.LaunchConfig;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.junit.Test;

/**
 * Test class for StreamSpecificAgent demonstrating streaming from a specific invocation.
 *
 * <p>This test demonstrates:
 *
 * <ul>
 *   <li>streamChunk: Emitting streaming data from agent nodes
 *   <li>agent.streamSpecific: Subscribing to streaming data from a specific invocation
 *   <li>agent.nextStep: Triggering execution after subscription
 *   <li>Selective monitoring of one invocation among multiple
 * </ul>
 */
public class StreamSpecificAgentTest {

  @Test
  public void testStreamSpecificAgentSingleInvocation() throws Exception {
    // Tests streaming from a specific agent invocation using streamSpecific
    try (InProcessCluster ipc = InProcessCluster.create()) {
      // Deploy the agent module
      StreamSpecificAgent.StreamSpecificAgentModule module =
          new StreamSpecificAgent.StreamSpecificAgentModule();
      ipc.launchModule(module, new LaunchConfig(1, 1));

      // Get agent manager and client
      String moduleName = module.getModuleName();
      AgentManager manager = AgentManager.create(ipc, moduleName);
      AgentClient agent = manager.getAgentClient("StreamSpecificAgent");

      // Track chunks received
      List<Map<String, Object>> chunksReceived = new ArrayList<>();

      // Start three invocations
      Map<String, Object> request1 = new HashMap<>();
      request1.put("taskId", "test-1");
      request1.put("itemsToProcess", 3);
      AgentInvoke invoke1 = agent.initiate(request1);

      Map<String, Object> request2 = new HashMap<>();
      request2.put("taskId", "test-2");
      request2.put("itemsToProcess", 4);
      AgentInvoke invoke2 = agent.initiate(request2);

      Map<String, Object> request3 = new HashMap<>();
      request3.put("taskId", "test-3");
      request3.put("itemsToProcess", 2);
      AgentInvoke invoke3 = agent.initiate(request3);

      // Get the invoke-id for the specific invocation to monitor
      UUID targetInvokeId = invoke2.getAgentInvokeId();

      // Subscribe to streaming from only invoke2
      agent.streamSpecific(
          invoke2,
          "process-task",
          targetInvokeId,
          (allChunks, newChunks, reset, complete) -> {
            for (Object chunkObj : newChunks) {
              @SuppressWarnings("unchecked")
              Map<String, Object> chunk = (Map<String, Object>) chunkObj;
              chunksReceived.add(chunk);
            }
          });

      // Trigger execution with nextStep
      agent.nextStep(invoke1);
      agent.nextStep(invoke2);
      agent.nextStep(invoke3);

      // Get final results
      @SuppressWarnings("unchecked")
      Map<String, Object> result1 = (Map<String, Object>) agent.result(invoke1);
      @SuppressWarnings("unchecked")
      Map<String, Object> result2 = (Map<String, Object>) agent.result(invoke2);
      @SuppressWarnings("unchecked")
      Map<String, Object> result3 = (Map<String, Object>) agent.result(invoke3);

      // Verify all results
      assertEquals("First task ID should be test-1", "test-1", result1.get("taskId"));
      assertEquals("First task should process 3 items", 3, (int) result1.get("totalItems"));
      assertEquals("First task should be completed", "completed", result1.get("status"));

      assertEquals("Second task ID should be test-2", "test-2", result2.get("taskId"));
      assertEquals("Second task should process 4 items", 4, (int) result2.get("totalItems"));
      assertEquals("Second task should be completed", "completed", result2.get("status"));

      assertEquals("Third task ID should be test-3", "test-3", result3.get("taskId"));
      assertEquals("Third task should process 2 items", 2, (int) result3.get("totalItems"));
      assertEquals("Third task should be completed", "completed", result3.get("status"));

      // Verify streaming chunks were received ONLY from invoke2
      assertEquals("Should receive 4 chunks from test-2 only", 4, chunksReceived.size());

      // Verify all chunks are from test-2
      for (Map<String, Object> chunk : chunksReceived) {
        assertEquals("All chunks should be from test-2", "test-2", chunk.get("taskId"));
        assertNotNull("Chunk should have itemNumber", chunk.get("itemNumber"));
        assertEquals("Chunk status should be processing", "processing", chunk.get("status"));
      }

      // Verify we got the expected item numbers
      List<Integer> itemNumbers = new ArrayList<>();
      for (Map<String, Object> chunk : chunksReceived) {
        itemNumbers.add((Integer) chunk.get("itemNumber"));
      }
      assertEquals("Should have item numbers 0-3", List.of(0, 1, 2, 3), itemNumbers);
    }
  }
}

File: examples/clj/test/com/rpl/agent/basic/stream_specific_agent_test.clj
(ns com.rpl.agent.basic.stream-specific-agent-test
  (:require
   [clojure.test :refer [deftest testing is]]
   [com.rpl.agent-o-rama :as aor]
   [com.rpl.rama :as rama]
   [com.rpl.rama.test :as rtest]
   [com.rpl.agent.basic.stream-specific-agent :refer [StreamSpecificAgentModule]]))

(deftest stream-specific-agent-test
  ;; Streaming from a specific invocation tests the ability to subscribe to
  ;; streaming chunks from one particular agent execution using its invoke-id,
  ;; verifying selective monitoring and agent-next-step behavior.
  (System/gc)
  (testing "StreamSpecificAgent"
    (with-open [ipc (rtest/create-ipc)]
      (rtest/launch-module! ipc StreamSpecificAgentModule {:tasks 1 :threads 1})

      (let [manager (aor/agent-manager ipc
                                       (rama/get-module-name
                                        StreamSpecificAgentModule))
            agent   (aor/agent-client manager "StreamSpecificAgent")]

        (testing "subscribes to streaming from a specific invocation only"
          (let [chunks-received (atom [])
                ;; Start three invocations
                invoke1 (aor/agent-initiate agent {:task-id "test-1" :items-to-process 3})
                invoke2 (aor/agent-initiate agent {:task-id "test-2" :items-to-process 4})
                invoke3 (aor/agent-initiate agent {:task-id "test-3" :items-to-process 2})
                ;; Get the invoke-id for the specific invocation to monitor
                target-invoke-id (:agent-invoke-id invoke2)]

            ;; Subscribe to streaming from only invoke2.
            ;; agent-stream-specific with agent-invoke-id targets this specific invocation.
            (aor/agent-stream-specific
             agent
             invoke2
             "process-task"
             target-invoke-id
             (fn [_all-chunks new-chunks _reset? _complete?]
               (doseq [chunk new-chunks]
                 (swap! chunks-received conj chunk))))

            ;; Get final results
            (let [result1 (aor/agent-result agent invoke1)
                  result2 (aor/agent-result agent invoke2)
                  result3 (aor/agent-result agent invoke3)]

              ;; Verify final results
              (is (= "test-1" (:task-id result1)))
              (is (= 3 (:total-items result1)))
              (is (= "completed" (:status result1)))

              (is (= "test-2" (:task-id result2)))
              (is (= 4 (:total-items result2)))
              (is (= "completed" (:status result2)))

              (is (= "test-3" (:task-id result3)))
              (is (= 2 (:total-items result3)))
              (is (= "completed" (:status result3)))

              ;; Verify streaming chunks were received ONLY from invoke2
              (is (= 4 (count @chunks-received)))
              (is (every? #(= "test-2" (:task-id %)) @chunks-received))

              ;; Verify chunk structure
              (is (every? #(contains? % :task-id) @chunks-received))
              (is (every? #(contains? % :item-number) @chunks-received))
              (is (every? #(= "processing" (:status %)) @chunks-received))

              ;; Verify we got the expected item numbers
              (is (= #{0 1 2 3} (set (map :item-number @chunks-received)))))))))))